bpf/lib/ipsecrps.h

```diff
@@ -0,0 +1,8 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Cilium */
+
+#pragma once
+
+/* Masks are in network byte order */
+#define IPSEC_RPS_SPI_MASK  0xFF000000
+#define IPSEC_RPS_HASH_MASK 0x00FFFFFF
```

Comment:
I was wondering about this :smile_cat:. This would make a fairly significant difference from a theoretical standpoint. I ran a simulation to investigate, where for every possible 24-bit and 8-bit hash, and for all CPU counts in the range of 1-255, I calculated the target CPU that a flow would be steered to based on its hash and kept a tally of how many flows a CPU would receive. The count was weighted based on the number of 32-bit hashes that the smaller hashes represent. For instance, if a CPU would receive a flow for the 8-bit hash 0xFF, then it would also receive flows for all the 32-bit hashes in the range 0x000000FF-0xFFFFFFFF. This allows us to simulate the distribution of flows across CPUs on a node and get a sense of how even or uneven it is, based on using 8-bit or 24-bit hashes.

The results were pretty stark and show that only using 8-bits could have a fairly large impact. This plot has the number of CPUs on the x-axis and the "variance-to-mean ratio" (mean-normalized variance) on the y-axis. The red bars indicate cases where the y-axis has a value of zero:

![Figure 12](https://github.com/user-attachments/assets/118a4069-9a05-40c1-a151-ba7af777f5ff)

The scale of the y-axis for the 8-bit hashes (purple) and 24-bit hashes (blue) reveals all. The mean-normalized variances for the 8-bit hashes are in the millions, while the mean-normalized variances for the 24-bit hashes are less than one. In cases where the number of CPUs on a node isn't optimal (in-between a power of 2), the distributions of flows is tolerable for 24-bit hashes and pretty bad for 8-bit hashes.

This graph is a plot of the difference between the most amount of flows a CPU received and the smallest number of flows a CPU received, for a given number of bits in the hash and the given number of simulated CPUs. The values aren't normalized, so it becomes clear to tell that, with 8-bits, there's an extra 15 million possible flows that a CPU could receive, and with 24-bits, there's only an extra 250.

![Figure 11](https://github.com/user-attachments/assets/839e6344-268f-4f27-b25e-e227f0cd6a43)

So, **TL;DR: using 8-bit hashes should be fine as long as you have a number of CPUs equal to a power of 2, otherwise it's up in the are and entirely dependent on your workload.** I'll run another performance benchmark to see what happens with 8-bits in a more practical sense though.